File: go/temporal/dto_test.go
package temporal

import (
	"encoding/json"
	"testing"

	core "github.com/SmooAI/smooth-operator-core/go/core"
)

// These tests exercise the serde boundary the durable path marshals across — the
// activity inputs and the model-call output — and run WITHOUT a Temporal server (no
// SDK runtime, just JSON), mirroring the Rust `dto.rs` unit tests that are always
// compiled regardless of the `temporal` feature.

// ModelCallInput survives a JSON round trip: the context window and tool specs cross
// the activity boundary intact.
func TestModelCallInput_SerdeRoundTrip(t *testing.T) {
	in := ModelCallInput{
		Messages: []core.ChatMessage{
			{Role: "system", Content: "you are a test agent"},
			{Role: "user", Content: "what is the durable answer?"},
		},
		Tools: []core.ToolSpec{
			{Name: "echo", Description: "echo it back", Parameters: map[string]any{"type": "object"}},
		},
	}
	blob, err := json.Marshal(in)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	var back ModelCallInput
	if err := json.Unmarshal(blob, &back); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if len(back.Messages) != 2 || back.Messages[1].Content != "what is the durable answer?" {
		t.Fatalf("messages not preserved: %+v", back.Messages)
	}
	if len(back.Tools) != 1 || back.Tools[0].Name != "echo" {
		t.Fatalf("tools not preserved: %+v", back.Tools)
	}
}

// ToolInvokeInput survives a JSON round trip: the tool call (id + name + raw-JSON
// arguments) crosses the activity boundary intact.
func TestToolInvokeInput_SerdeRoundTrip(t *testing.T) {
	in := ToolInvokeInput{Call: core.ToolCall{ID: "call-1", Name: "echo", Arguments: `{"text":"hi"}`}}
	blob, err := json.Marshal(in)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	var back ToolInvokeInput
	if err := json.Unmarshal(blob, &back); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if back.Call.ID != "call-1" || back.Call.Name != "echo" || back.Call.Arguments != `{"text":"hi"}` {
		t.Fatalf("call not preserved: %+v", back.Call)
	}
}

// The model-call activity output — core.ChatResponse used directly as the DTO —
// survives a JSON round trip preserving every field the orchestration and
// accounting read (content, tool calls, usage, gateway cost). This is the Go analog
// of the Rust `model_call_output_round_trips_through_llm_response` test; Go needs no
// separate projection type because ChatResponse is already fully serializable.
func TestChatResponse_ActivityOutputRoundTrip(t *testing.T) {
	cost := 0.0001
	original := core.ChatResponse{
		Content: "the durable answer is 42",
		ToolCalls: []core.ToolCall{
			{ID: "call-1", Name: "echo", Arguments: `{"text":"hi"}`},
		},
		Usage:          core.Usage{PromptTokens: 10, CompletionTokens: 5},
		GatewayCostUSD: &cost,
	}
	blob, err := json.Marshal(original)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	var back core.ChatResponse
	if err := json.Unmarshal(blob, &back); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if back.Content != original.Content {
		t.Errorf("content: got %q", back.Content)
	}
	if len(back.ToolCalls) != 1 || back.ToolCalls[0].ID != "call-1" {
		t.Errorf("tool calls not preserved: %+v", back.ToolCalls)
	}
	if back.Usage.TotalTokens() != 15 {
		t.Errorf("usage total: got %d", back.Usage.TotalTokens())
	}
	if back.GatewayCostUSD == nil || *back.GatewayCostUSD != cost {
		t.Errorf("gateway cost not preserved: %v", back.GatewayCostUSD)
	}
}

// parseWaitSeconds accepts a whole non-negative integer and rejects everything else
// (missing, non-numeric, negative, fractional, unparseable) — the guard the durable
// wait tool relies on to fail cleanly instead of sleeping on garbage.
func TestParseWaitSeconds(t *testing.T) {
	cases := []struct {
		args   string
		want   int64
		wantOK bool
	}{
		{`{"seconds":1}`, 1, true},
		{`{"seconds":0}`, 0, true},
		{`{"seconds":86400}`, 86400, true},
		{`{}`, 0, false},
		{``, 0, false},
		{`{"seconds":"5"}`, 0, false},
		{`{"seconds":-3}`, 0, false},
		{`{"seconds":1.5}`, 0, false},
		{`not json`, 0, false},
	}
	for _, c := range cases {
		got, ok := parseWaitSeconds(c.args)
		if ok != c.wantOK || got != c.want {
			t.Errorf("parseWaitSeconds(%q) = (%d, %v), want (%d, %v)", c.args, got, ok, c.want, c.wantOK)
		}
	}
}

File: typescript/temporal/test/dto.test.ts
/**
 * Unit tests for the activity DTO boundary — the TS sibling of the Rust crate's
 * `dto.rs` tests. These run with **no Temporal runtime**: they prove the model
 * response projects to a JSON-serializable DTO and reconstructs the fields the
 * deterministic `driveTurn` loop reads, i.e. that it can actually cross an
 * activity boundary.
 */

import { describe, expect, it } from 'vitest';

import type { ModelResponse } from '@smooai/smooth-operator-core/executor';

import { modelResponseToOutput, outputToModelResponse, type ModelCallInput, type ToolInvokeInput } from '../src/dto.js';

function sampleResponse(): ModelResponse {
    return {
        choices: [
            {
                message: {
                    content: 'hello',
                    tool_calls: [{ id: 'call-1', function: { name: 'echo', arguments: '{"text":"hi"}' } }],
                },
            },
        ],
        usage: { prompt_tokens: 10, completion_tokens: 5 },
    } as ModelResponse;
}

describe('ModelCallOutput', () => {
    it('projects a ModelResponse then reconstructs the fields the loop reads', () => {
        const dto = modelResponseToOutput(sampleResponse());
        expect(dto.content).toBe('hello');
        expect(dto.toolCalls).toHaveLength(1);
        expect(dto.toolCalls[0].id).toBe('call-1');
        expect(dto.toolCalls[0].function.name).toBe('echo');
        expect(dto.usage).toEqual({ prompt_tokens: 10, completion_tokens: 5 });

        const restored = outputToModelResponse(dto);
        expect(restored.choices[0].message.content).toBe('hello');
        expect(restored.choices[0].message.tool_calls).toHaveLength(1);
        expect(restored.choices[0].message.tool_calls?.[0].function.arguments).toBe('{"text":"hi"}');
        expect(restored.usage).toEqual({ prompt_tokens: 10, completion_tokens: 5 });
    });

    it('survives a JSON serialize/deserialize round trip (crosses an activity boundary)', () => {
        const dto = modelResponseToOutput(sampleResponse());
        const back = JSON.parse(JSON.stringify(dto)) as typeof dto;
        expect(back.content).toBe('hello');
        expect(back.toolCalls[0].function.name).toBe('echo');
        expect(back.usage?.prompt_tokens).toBe(10);
    });

    it('a tool-only reply (no content) projects to empty content + null tool_calls when reconstructed empty', () => {
        const toolOnly = {
            choices: [{ message: { content: null, tool_calls: [{ id: 'c1', function: { name: 't', arguments: '{}' } }] } }],
            usage: null,
        } as ModelResponse;
        const dto = modelResponseToOutput(toolOnly);
        expect(dto.content).toBe('');
        expect(dto.toolCalls).toHaveLength(1);
        expect(dto.usage).toBeUndefined();

        // A response with no tool calls reconstructs to null tool_calls (OpenAI wire shape).
        const noTools = outputToModelResponse({ content: 'done', toolCalls: [] });
        expect(noTools.choices[0].message.tool_calls).toBeNull();
    });
});

describe('activity inputs', () => {
    it('serialize cleanly (JSON round trip)', () => {
        const modelInput: ModelCallInput = {
            messages: [{ role: 'user', content: 'hi' }],
            tools: [{ type: 'function', function: { name: 'echo', description: 'echo', parameters: { type: 'object' } } }],
        };
        const backModel = JSON.parse(JSON.stringify(modelInput)) as ModelCallInput;
        expect(backModel.messages).toHaveLength(1);
        expect((backModel.tools[0].function as { name: string }).name).toBe('echo');

        const toolInput: ToolInvokeInput = { call: { id: 'c1', name: 'echo', arguments: {} } };
        const backTool = JSON.parse(JSON.stringify(toolInput)) as ToolInvokeInput;
        expect(backTool.call.name).toBe('echo');
    });
});
